File: frontend/test/value_normalization_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:provaluer_frontend/features/document_workspace/services/alias_resolution_engine.dart';
import 'package:provaluer_frontend/features/document_workspace/services/value_normalization_engine.dart';

void main() {
  group('Value Normalization Engine (Dart)', () {
    test('Unit-Aware Area Normalization correctly derives raw, numeric, detected unit, and standard sqft', () {
      // 2.375 Acres
      final acreVal = ValueNormalizationEngine.normalizeAreaValue('2.375 Acres');
      expect(acreVal.rawValue, equals('2.375 Acres'));
      expect(acreVal.numericValue, equals(2.375));
      expect(acreVal.detectedUnit, equals('ACRES'));
      expect(acreVal.standardSqftValue, equals(103455.0));

      // 250 Sq.Yd
      final sqydVal = ValueNormalizationEngine.normalizeAreaValue('250 Sq.Yd');
      expect(sqydVal.rawValue, equals('250 Sq.Yd'));
      expect(sqydVal.numericValue, equals(250.0));
      expect(sqydVal.detectedUnit, equals('SQYD'));
      expect(sqydVal.standardSqftValue, equals(2250.0));

      // 100 Sq.M
      final sqmVal = ValueNormalizationEngine.normalizeAreaValue('100 Sq.M');
      expect(sqmVal.rawValue, equals('100 Sq.M'));
      expect(sqmVal.numericValue, equals(100.0));
      expect(sqmVal.detectedUnit, equals('SQM'));
      expect(sqmVal.standardSqftValue, equals(1076.391));

      // 40 Cents
      final centsVal = ValueNormalizationEngine.normalizeAreaValue('40 Cents');
      expect(centsVal.rawValue, equals('40 Cents'));
      expect(centsVal.numericValue, equals(40.0));
      expect(centsVal.detectedUnit, equals('CENTS'));
      expect(centsVal.standardSqftValue, equals(17424.0));

      // 5 Grounds
      final groundsVal = ValueNormalizationEngine.normalizeAreaValue('5 Grounds');
      expect(groundsVal.rawValue, equals('5 Grounds'));
      expect(groundsVal.numericValue, equals(5.0));
      expect(groundsVal.detectedUnit, equals('GROUNDS'));
      expect(groundsVal.standardSqftValue, equals(12000.0));

      // 1000.125 Sq.Ft
      final sqftVal = ValueNormalizationEngine.normalizeAreaValue('1000.125 Sq.Ft');
      expect(sqftVal.rawValue, equals('1000.125 Sq.Ft'));
      expect(sqftVal.numericValue, equals(1000.125));
      expect(sqftVal.detectedUnit, equals('SQFT'));
      expect(sqftVal.standardSqftValue, equals(1000.125));
    });

    test('Area Normalization extracts exact numeric values up to 3 decimal places', () {
      expect(ValueNormalizationEngine.normalize('SALEABLE_AREA', '1000 sq.ft'), equals(1000.0));
      expect(ValueNormalizationEngine.normalize('SALEABLE_AREA', '1000.125 sq.ft'), equals(1000.125));
      expect(ValueNormalizationEngine.normalize('SUPER_BUILT_UP_AREA', '1000 sqft'), equals(1000.0));
      expect(ValueNormalizationEngine.normalize('PROPERTY_AREA_SFT', '1000 SFT'), equals(1000.0));
      expect(ValueNormalizationEngine.normalize('SBUA', '1000 Square Feet'), equals(1000.0));
      expect(ValueNormalizationEngine.normalize('FLAT_AREA', 'Approx 1000 Sq.Ft'), equals(1000.0));
      expect(ValueNormalizationEngine.normalize('SALEABLE_AREA', 'Approx. 1000 Sq.Ft'), equals(1000.0));
      expect(ValueNormalizationEngine.normalize('SALEABLE_AREA', 'Area = 1000 Sq Ft'), equals(1000.0));
      expect(ValueNormalizationEngine.normalize('SALEABLE_AREA', '1,250.50 sq.ft'), equals(1250.5));
      expect(ValueNormalizationEngine.normalize('LAND_AREA', '2.5 Acres'), equals(2.5));
      expect(ValueNormalizationEngine.normalize('LAND_AREA', '2.375 Acres'), equals(2.375));
      expect(ValueNormalizationEngine.normalize('PLOT_AREA', '5.0 Grounds'), equals(5.0));
      expect(ValueNormalizationEngine.normalize('PLOT_AREA', '5.125 Grounds'), equals(5.125));
    });

    test('Rate Normalization extracts exact numeric rates', () {
      expect(ValueNormalizationEngine.normalize('MARKET_RATE_FLAT', 'Rs 5000'), equals(5000.0));
      expect(ValueNormalizationEngine.normalize('MARKET_RATE_FLAT', 'Rs. 5000'), equals(5000.0));
      expect(ValueNormalizationEngine.normalize('MARKET_RATE_FLAT', 'Rs 5000.125'), equals(5000.125));
      expect(ValueNormalizationEngine.normalize('COMPOSITE_RATE', 'INR 6000'), equals(6000.0));
      expect(ValueNormalizationEngine.normalize('CURRENT_MARKET_RATE', '₹5000'), equals(5000.0));
      expect(ValueNormalizationEngine.normalize('FLAT_MARKET_RATE', '₹ 7,500'), equals(7500.0));
      expect(ValueNormalizationEngine.normalize('BUILDING_MARKET_RATE', '₹ 7,500 per sq.ft'), equals(7500.0));
      expect(ValueNormalizationEngine.normalize('MARKET_RATE_FLAT', '₹ 7500.750 per sq.ft'), equals(7500.75));
      expect(ValueNormalizationEngine.normalize('LAND_RATE', 'Rs. 8,500/SFT'), equals(8500.0));
      expect(ValueNormalizationEngine.normalize('GOVERNMENT_RATE', '5000/-'), equals(5000.0));
    });

    test('Percentage Normalization handles percentages, decimals, and negative values', () {
      expect(ValueNormalizationEngine.normalize('REALIZABLE_PERCENTAGE', '85%'), equals(85.0));
      expect(ValueNormalizationEngine.normalize('REALIZABLE_PERCENTAGE', '90 %'), equals(90.0));
      expect(ValueNormalizationEngine.normalize('DISTRESS_SALE_PERCENTAGE', '12.5%'), equals(12.5));
      expect(ValueNormalizationEngine.normalize('DISTRESS_SALE_PERCENTAGE', '12.875%'), equals(12.875));
      expect(ValueNormalizationEngine.normalize('COMPOSITE_BUILDING_DEPRECIATION_PCT', '-10%'), equals(-10.0));
    });

    test('Decimal Precision Governance preserves up to 3 decimals and rounds > 3 with HALF_UP', () {
      expect(ValueNormalizationEngine.normalize('SALEABLE_AREA', '1000.1234 sq.ft'), equals(1000.123));
      expect(ValueNormalizationEngine.normalize('SALEABLE_AREA', '1000.56789 sq.ft'), equals(1000.568));
      expect(ValueNormalizationEngine.normalize('LAND_AREA', '2.9999 Acres'), equals(3.0));
    });

    test('Validation Rejections: Non-numeric inputs reject with expected error message', () {
      expect(
        () => ValueNormalizationEngine.normalize('SALEABLE_AREA', 'abc'),
        throwsA(isA<FormatException>().having((e) => e.message, 'message', ValueNormalizationEngine.validationErrorMsg)),
      );
      expect(
        () => ValueNormalizationEngine.normalize('MARKET_RATE_FLAT', 'Rs only'),
        throwsA(isA<FormatException>().having((e) => e.message, 'message', ValueNormalizationEngine.validationErrorMsg)),
      );
      expect(
        () => ValueNormalizationEngine.normalize('SUPER_BUILT_UP_AREA', 'sq.ft only'),
        throwsA(isA<FormatException>().having((e) => e.message, 'message', ValueNormalizationEngine.validationErrorMsg)),
      );
      expect(
        () => ValueNormalizationEngine.normalize('REALIZABLE_PERCENTAGE', ''),
        throwsA(isA<FormatException>().having((e) => e.message, 'message', ValueNormalizationEngine.validationErrorMsg)),
      );
    });

    test('Calculation Accuracy: Arithmetic preserves 100% calculation accuracy', () {
      final area1 = ValueNormalizationEngine.normalize('SALEABLE_AREA', '1000 sq.ft');
      final rate1 = ValueNormalizationEngine.normalize('MARKET_RATE_FLAT', 'Rs 5000');
      expect(area1 * rate1, equals(5000000.0));

      final area2 = ValueNormalizationEngine.normalize('SUPER_BUILT_UP_AREA', '1200 sqft');
      final rate2 = ValueNormalizationEngine.normalize('COMPOSITE_RATE', 'INR 6000');
      expect(area2 * rate2, equals(7200000.0));

      final area3 = ValueNormalizationEngine.normalize('SALEABLE_AREA', '1000.125 sq.ft');
      final rate3 = ValueNormalizationEngine.normalize('MARKET_RATE_FLAT', 'Rs 5000.250');
      expect(area3 * rate3, equals(5000875.03125));
    });

    test('Authoritative Area Calculation: Uses Standard Area in Sq.Ft', () {
      // 2.375 Acres @ Rs 5000/Sq.Ft
      final acreVal = ValueNormalizationEngine.normalizeAreaValue('2.375 Acres');
      final rate = ValueNormalizationEngine.normalizeRate('Rs 5000');
      final unitAmount = acreVal.standardSqftValue * rate;
      expect(unitAmount, equals(517275000.0)); // 103455 * 5000
    });

    test('Alias Resolution maps canonical keys and preserves suffixes', () {
      expect(AliasResolutionEngine.resolveCanonical('SUPER_BUILT_UP_AREA'), equals('SALEABLE_AREA'));
      expect(AliasResolutionEngine.resolveCanonical('PROPERTY_AREA_SFT'), equals('SALEABLE_AREA'));
      expect(AliasResolutionEngine.resolveCanonical('SBUA'), equals('SALEABLE_AREA'));
      expect(AliasResolutionEngine.resolveCanonical('FLAT_AREA'), equals('SALEABLE_AREA'));
      expect(AliasResolutionEngine.resolveCanonical('SALEABLE_AREA_SQFT'), equals('SALEABLE_AREA'));

      expect(AliasResolutionEngine.resolveCanonical('COMPOSITE_RATE'), equals('MARKET_RATE_FLAT'));
      expect(AliasResolutionEngine.resolveCanonical('CURRENT_MARKET_RATE'), equals('MARKET_RATE_FLAT'));
      expect(AliasResolutionEngine.resolveCanonical('FLAT_MARKET_RATE'), equals('MARKET_RATE_FLAT'));
      expect(AliasResolutionEngine.resolveCanonical('BUILDING_MARKET_RATE'), equals('MARKET_RATE_FLAT'));

      expect(AliasResolutionEngine.resolveCanonical('SUPER_BUILT_UP_AREA_RAW'), equals('SALEABLE_AREA_RAW'));
      expect(AliasResolutionEngine.resolveCanonical('SBUA_NUMERIC'), equals('SALEABLE_AREA_NUMERIC'));
    });
  });
}

File: backend/src/test/java/com/provaluer/service/ValueNormalizationEngineTest.java
package com.provaluer.service;

import com.provaluer.util.ValueNormalizationEngine;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

import static org.junit.jupiter.api.Assertions.*;

public class ValueNormalizationEngineTest {

    @Test
    @DisplayName("Unit-Aware Area Normalization correctly derives raw, numeric, detected unit, and standard sqft")
    public void testUnitAwareAreaNormalization() {
        // 2.375 Acres
        ValueNormalizationEngine.NormalizedValue acreVal = ValueNormalizationEngine.normalizeAreaValue("2.375 Acres");
        assertEquals("2.375 Acres", acreVal.getRawValue());
        assertEquals(0, new BigDecimal("2.375").compareTo(acreVal.getNumericValue()));
        assertEquals("ACRES", acreVal.getDetectedUnit());
        assertEquals(0, new BigDecimal("103455.000").compareTo(acreVal.getStandardSqftValue()));

        // 250 Sq.Yd
        ValueNormalizationEngine.NormalizedValue sqydVal = ValueNormalizationEngine.normalizeAreaValue("250 Sq.Yd");
        assertEquals("250 Sq.Yd", sqydVal.getRawValue());
        assertEquals(0, new BigDecimal("250").compareTo(sqydVal.getNumericValue()));
        assertEquals("SQYD", sqydVal.getDetectedUnit());
        assertEquals(0, new BigDecimal("2250.000").compareTo(sqydVal.getStandardSqftValue()));

        // 100 Sq.M
        ValueNormalizationEngine.NormalizedValue sqmVal = ValueNormalizationEngine.normalizeAreaValue("100 Sq.M");
        assertEquals("100 Sq.M", sqmVal.getRawValue());
        assertEquals(0, new BigDecimal("100").compareTo(sqmVal.getNumericValue()));
        assertEquals("SQM", sqmVal.getDetectedUnit());
        assertEquals(0, new BigDecimal("1076.391").compareTo(sqmVal.getStandardSqftValue()));

        // 40 Cents
        ValueNormalizationEngine.NormalizedValue centsVal = ValueNormalizationEngine.normalizeAreaValue("40 Cents");
        assertEquals("40 Cents", centsVal.getRawValue());
        assertEquals(0, new BigDecimal("40").compareTo(centsVal.getNumericValue()));
        assertEquals("CENTS", centsVal.getDetectedUnit());
        assertEquals(0, new BigDecimal("17424.000").compareTo(centsVal.getStandardSqftValue()));

        // 5 Grounds
        ValueNormalizationEngine.NormalizedValue groundsVal = ValueNormalizationEngine.normalizeAreaValue("5 Grounds");
        assertEquals("5 Grounds", groundsVal.getRawValue());
        assertEquals(0, new BigDecimal("5").compareTo(groundsVal.getNumericValue()));
        assertEquals("GROUNDS", groundsVal.getDetectedUnit());
        assertEquals(0, new BigDecimal("12000.000").compareTo(groundsVal.getStandardSqftValue()));

        // 1000.125 Sq.Ft
        ValueNormalizationEngine.NormalizedValue sqftVal = ValueNormalizationEngine.normalizeAreaValue("1000.125 Sq.Ft");
        assertEquals("1000.125 Sq.Ft", sqftVal.getRawValue());
        assertEquals(0, new BigDecimal("1000.125").compareTo(sqftVal.getNumericValue()));
        assertEquals("SQFT", sqftVal.getDetectedUnit());
        assertEquals(0, new BigDecimal("1000.125").compareTo(sqftVal.getStandardSqftValue()));
    }

    @Test
    @DisplayName("Area Normalization extracts exact numeric values ignoring prefixes, units, and punctuation")
    public void testAreaNormalization() {
        assertEquals(0, new BigDecimal("1000").compareTo(ValueNormalizationEngine.normalizeArea("1000 sq.ft")));
        assertEquals(0, new BigDecimal("1000.125").compareTo(ValueNormalizationEngine.normalizeArea("1000.125 sq.ft")));
        assertEquals(0, new BigDecimal("1000").compareTo(ValueNormalizationEngine.normalizeArea("1000 sqft")));
        assertEquals(0, new BigDecimal("1000").compareTo(ValueNormalizationEngine.normalizeArea("1000 SFT")));
        assertEquals(0, new BigDecimal("1000").compareTo(ValueNormalizationEngine.normalizeArea("1000 Square Feet")));
        assertEquals(0, new BigDecimal("1000").compareTo(ValueNormalizationEngine.normalizeArea("Approx 1000 Sq.Ft")));
        assertEquals(0, new BigDecimal("1000").compareTo(ValueNormalizationEngine.normalizeArea("Approx. 1000 Sq.Ft")));
        assertEquals(0, new BigDecimal("1000").compareTo(ValueNormalizationEngine.normalizeArea("Area = 1000 Sq Ft")));
        assertEquals(0, new BigDecimal("1250.50").compareTo(ValueNormalizationEngine.normalizeArea("1,250.50 sq.ft")));
        assertEquals(0, new BigDecimal("2.5").compareTo(ValueNormalizationEngine.normalizeArea("2.5 Acres")));
        assertEquals(0, new BigDecimal("2.375").compareTo(ValueNormalizationEngine.normalizeArea("2.375 Acres")));
        assertEquals(0, new BigDecimal("5.0").compareTo(ValueNormalizationEngine.normalizeArea("5.0 Grounds")));
        assertEquals(0, new BigDecimal("5.125").compareTo(ValueNormalizationEngine.normalizeArea("5.125 Grounds")));
    }

    @Test
    @DisplayName("Rate Normalization extracts exact numeric rates ignoring currency symbols, slashes, and qualifiers")
    public void testRateNormalization() {
        assertEquals(0, new BigDecimal("5000").compareTo(ValueNormalizationEngine.normalizeRate("Rs 5000")));
        assertEquals(0, new BigDecimal("5000").compareTo(ValueNormalizationEngine.normalizeRate("Rs. 5000")));
        assertEquals(0, new BigDecimal("5000.125").compareTo(ValueNormalizationEngine.normalizeRate("Rs 5000.125")));
        assertEquals(0, new BigDecimal("6000").compareTo(ValueNormalizationEngine.normalizeRate("INR 6000")));
        assertEquals(0, new BigDecimal("5000").compareTo(ValueNormalizationEngine.normalizeRate("₹5000")));
        assertEquals(0, new BigDecimal("7500").compareTo(ValueNormalizationEngine.normalizeRate("₹ 7,500")));
        assertEquals(0, new BigDecimal("7500").compareTo(ValueNormalizationEngine.normalizeRate("₹ 7,500 per sq.ft")));
        assertEquals(0, new BigDecimal("7500.750").compareTo(ValueNormalizationEngine.normalizeRate("₹ 7500.750 per sq.ft")));
        assertEquals(0, new BigDecimal("8500").compareTo(ValueNormalizationEngine.normalizeRate("Rs. 8,500/SFT")));
        assertEquals(0, new BigDecimal("5000").compareTo(ValueNormalizationEngine.normalizeRate("5000/-")));
    }

    @Test
    @DisplayName("Percentage Normalization handles percentages, decimals, and negative values")
    public void testPercentageNormalization() {
        assertEquals(0, new BigDecimal("85").compareTo(ValueNormalizationEngine.normalizePercentage("85%")));
        assertEquals(0, new BigDecimal("90").compareTo(ValueNormalizationEngine.normalizePercentage("90 %")));
        assertEquals(0, new BigDecimal("12.5").compareTo(ValueNormalizationEngine.normalizePercentage("12.5%")));
        assertEquals(0, new BigDecimal("12.875").compareTo(ValueNormalizationEngine.normalizePercentage("12.875%")));
        assertEquals(0, new BigDecimal("-10").compareTo(ValueNormalizationEngine.normalizePercentage("-10%")));
    }

    @Test
    @DisplayName("Decimal Precision Governance preserves up to 3 decimal places and rounds > 3 to 3 with HALF_UP")
    public void testDecimalPrecisionGovernance() {
        assertEquals("1000.123", ValueNormalizationEngine.normalize("SALEABLE_AREA", "1000.1234 sq.ft").toPlainString());
        assertEquals("1000.568", ValueNormalizationEngine.normalize("SALEABLE_AREA", "1000.56789 sq.ft").toPlainString());
        assertEquals("3.000", ValueNormalizationEngine.normalize("LAND_AREA", "2.9999 Acres").toPlainString());
        assertEquals("1000.125", ValueNormalizationEngine.normalize("SALEABLE_AREA", "1000.125 sq.ft").toPlainString());
        assertEquals("5000.25", ValueNormalizationEngine.normalize("MARKET_RATE_FLAT", "Rs 5000.25").toPlainString());
    }

    @Test
    @DisplayName("Validation Rejections: Non-numeric inputs reject with expected error message and no silent conversion to zero")
    public void testRejections() {
        IllegalArgumentException ex1 = assertThrows(IllegalArgumentException.class, () ->
                ValueNormalizationEngine.normalize("SALEABLE_AREA", "abc"));
        assertEquals("Numeric value could not be derived from input.", ex1.getMessage());

        IllegalArgumentException ex2 = assertThrows(IllegalArgumentException.class, () ->
                ValueNormalizationEngine.normalize("MARKET_RATE_FLAT", "Rs only"));
        assertEquals("Numeric value could not be derived from input.", ex2.getMessage());

        IllegalArgumentException ex3 = assertThrows(IllegalArgumentException.class, () ->
                ValueNormalizationEngine.normalize("SUPER_BUILT_UP_AREA", "sq.ft only"));
        assertEquals("Numeric value could not be derived from input.", ex3.getMessage());

        IllegalArgumentException ex4 = assertThrows(IllegalArgumentException.class, () ->
                ValueNormalizationEngine.normalize("REALIZABLE_PERCENTAGE", ""));
        assertEquals("Numeric value could not be derived from input.", ex4.getMessage());
    }

    @Test
    @DisplayName("Calculation Accuracy: Authoritative Area Calculation uses STANDARD_AREA_SQFT")
    public void testAuthoritativeAreaCalculation() {
        // 2.375 Acres @ Rs 5000/sq.ft
        ValueNormalizationEngine.NormalizedValue acreArea = ValueNormalizationEngine.normalizeAreaValue("2.375 Acres");
        BigDecimal rate = ValueNormalizationEngine.normalizeRate("Rs 5000");

        // Authoritative calculation rule: UNIT_AMOUNT = STANDARD_AREA_SQFT * MARKET_RATE_FLAT_NUMERIC
        BigDecimal unitAmount = acreArea.getStandardSqftValue().multiply(rate);
        assertEquals(0, new BigDecimal("517275000").compareTo(unitAmount)); // 103455 * 5000 = 517,275,000

        // 1000 sq.ft @ Rs 5000
        ValueNormalizationEngine.NormalizedValue sqftArea = ValueNormalizationEngine.normalizeAreaValue("1000 sq.ft");
        assertEquals(0, new BigDecimal("5000000").compareTo(sqftArea.getStandardSqftValue().multiply(rate)));
    }
}
